File: src/tut01_nodes/solution/tut01_nodes_solution/rate_listener.py
# Copyright 2026 Jose Laruta
"""Exercise 2: measure the rate of `chatter` and log it every 5 seconds."""

import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from std_msgs.msg import String


class RateListener(Node):
    """Counts messages in a subscription callback, reports from a timer callback."""

    def __init__(self) -> None:
        super().__init__('rate_listener')
        self._count = 0
        self._window_start = self.get_clock().now()
        # Both callbacks run on the same thread (single-threaded executor), so
        # the counter needs no lock. Lesson 6 revisits this.
        self._sub = self.create_subscription(String, 'chatter', self._on_message, 10)
        self._timer = self.create_timer(5.0, self._on_timer)
        self.get_logger().info('rate_listener started')

    def _on_message(self, msg: String) -> None:
        """Count one message; the content does not matter here."""
        self._count += 1

    def _on_timer(self) -> None:
        """Log the rate over the last window and start a new one."""
        window_end = self.get_clock().now()
        elapsed = (window_end - self._window_start).nanoseconds / 1e9
        rate = self._count / elapsed if elapsed > 0.0 else 0.0
        self.get_logger().info(f'{self._count} messages in {elapsed:.1f} s: {rate:.2f} Hz')
        self._count = 0
        self._window_start = window_end


def main(args: list[str] | None = None) -> None:
    """Entry point: init, spin, and shut down cleanly."""
    rclpy.init(args=args)
    node = RateListener()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == '__main__':
    main()

File: src/tut01_nodes/solution/src/rate_listener.cpp
// Copyright 2026 Jose Laruta

#include <chrono>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

/// Counts `chatter` messages and logs the measured rate every 5 seconds.
class RateListener : public rclcpp::Node
{
public:
  RateListener()
  : Node("rate_listener")
  {
    window_start_ = now();
    // Both callbacks run on the same thread (single-threaded executor), so the
    // counter needs no lock. Lesson 6 revisits this.
    sub_ = create_subscription<std_msgs::msg::String>(
      "chatter", 10, [this](const std_msgs::msg::String &) {++count_;});
    timer_ = create_wall_timer(5s, [this]() {on_timer();});
    RCLCPP_INFO(get_logger(), "rate_listener started");
  }

private:
  /// Log the rate over the last window and start a new one.
  void on_timer()
  {
    const auto window_end = now();
    const double elapsed = (window_end - window_start_).seconds();
    const double rate = elapsed > 0.0 ? count_ / elapsed : 0.0;
    RCLCPP_INFO(get_logger(), "%d messages in %.1f s: %.2f Hz", count_, elapsed, rate);
    count_ = 0;
    window_start_ = window_end;
  }

  int count_{0};
  rclcpp::Time window_start_;
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<RateListener>());
  rclcpp::shutdown();
  return 0;
}
